Example: News Classification with GPT-2 (#24)
opst/knitfab-docs: 04.examples/news-classification/news-classification.ja.md

### 例
# Knitfab での LLM のファインチューニングによるニュース分類

> 本書では、Knitfab を使用した一般大規模言語モデル (LLM) をファインチューニングすることにより、モデルがニュースといった特定の文書分類性能を向上することを示します。この事例を通して、Knitfab がファインチューニング実験の入力と出力、および履歴の管理を自動化しながら、LLM-as-a-judge を用いて効率的に人間のような評価ができるようになります。

Reviewer: ```suggestion
本書では、Knitfab を使用した一般大規模言語モデル (LLM) をファインチューニングすることにより、モデルのニュースといった特定の文書分類性能を改善できることを示します。この事例を通して、Knitfab がファインチューニング実験の入力と出力、および履歴の管理を自動化しながら、LLM-as-a-judge を用いて効率的に人間のような評価ができるようになります。
```

### リポジトリ
本書で使用されるファイルとディレクトリは、GitHub の `knitfab-docs` リポジトリからをクローンできます。
```bash
git clone https://github.com/opst/knitfab-docs.git
```
クローンが完了したら、`04.examples/news-classification` ディレクトリ内にあるファイルを確認してください。
- **scripts:** このディレクトリには、モデルの学習と登録に必要なファイルが含まれています。
  - モデルの学習と評価のための Python スクリプト (`train.py`、`evaluate.py`)。
  - 必要なすべての Python パッケージを管理する `requirements.txt`。
  - Knitfab Plan の登録に使う Docker イメージを構築するための `Dockerfile`。
- **configs:** 学習と評価タスクの構成情報が含まれています。
- **plans:** Knitfab Plan の YAML ひな型が含まれています。

### タスク
- [ステップ 1: ニュース分類のファインチューニングタスクの定義](#ステップ-1-ニュース分類のファインチューニングタスクの定義)
- [ステップ 2: LLM-as-a-judge 評価タスクの定義](#ステップ-2-LLM-as-a-judge-評価タスクの定義)
- [ステップ 3: Docker イメージのビルド](#ステップ-3-docker-イメージのビルド)
- [ステップ 4:（任意）Docker イメージの動作確認](#ステップ-4任意docker-イメージの動作確認)
- [ステップ 5: Docker イメージを Knitfab にプッシュする](#ステップ-5-docker-イメージを-Knitfab-にプッシュする)
- [ステップ 6: ニュース分類のファインチューニング](#ステップ-6-ニュース分類のファインチューニング)
- [ステップ 7: LLM-as-a-judge 評価](#ステップ-7-LLM-as-a-judge-評価)
- [ステップ 8: 片付け](#ステップ-8-片付け)

## ステップ 1: ニュース分類のファインチューニングタスクの定義

本章では、`scripts/train/train.py` にニュース分類ファインチューニングのロジックを説明します。スクリプトは引数指定、モデル準備、データ処理、学習と、評価を構造化したパイプラインに従い、各コンポーネントの構成変更と最適化を容易にする設計を取っています。

**1. 引数指定**

> `parse_arguments` 関数は、学習タスクに必要な引数を受け取ります。さまざまなパラメータを任意に指定できるため、モデルのファインチューニングに柔軟に変更できます。

Reviewer: ```suggestion
`parse_arguments` 関数は、学習タスクに必要な引数を受け取ります。さまざまなパラメータを任意に指定できるため、モデルのファインチューニングに柔軟に対応できます。
```

    per_device_eval_batch_size=1,
    gradient_accumulation_steps=2,
    learning_rate=self.args.learning_rate,
    weight_decay=0.05,
    max_grad_norm=0.3,
    fp16=True,
    bf16=False,
    warmup_ratio=0.1,
    lr_scheduler_type="cosine",
    logging_strategy="steps",
)
```
- バッチサイズ、学習率、重み減衰、勾配累積など、主要な学習ハイパーパラメータを定義します。
- 学習を効率化するために、混合精度 (`fp16`) を実装します。
- ウォームアップ比率 `0.1` 、とコサイン学習率スケジューラを使用します。
- 学習済みモデルやログファイルなど、プロジェクトの出力を整理するための専用ディレクトリを指定します。

```python
peft_config = LoraConfig(
    lora_alpha=16,
    lora_dropout=0.1,
    r=64,
    bias="none",
    task_type="SEQ_CLS",
)
```
- 効率的なファインチューニングのために `LoRA (Low-Rank Adaptation)` を採用します。
- 正則化とモデル分類性能の均衡を保つために、ドロップアウトとランク値を設定します。

```python
collator = DataCollatorWithPadding(tokenizer=self.tokenizer, padding="max_length")
```
- 文章を同じ固定長に変換します。

```python
self.trainer = SFTTrainer(
    model=self.model,
    args=training_args,
    peft_config=peft_config,
    train_dataset=self.train_dataset,
    eval_dataset=self.eval_dataset,
    tokenizer=self.tokenizer,
    data_collator=collator,
)
```
- 構成済みのモデル、学習引数、データセット、およびトークナイザーを使用して `SFTTrainer` を初期化します。
- すべての要素を組み合わせて、分類ファインチューニングを効率的に実行します。

**5. モデルの学習と評価**

> `run` 関数は、環境設定から結果の保存まで、学習プロセス全体を調整します。

Reviewer: ```suggestion
`run` 関数は、環境設定から結果の保存まで、学習プロセス全体を順次実行します。
```

- データセットを読み込み、学習器を初期化します。
- 学習と評価を実装します。
- 学習済みモデルと評価指標を保存します。

**コード解析**
```python
self.trainer.train()
self.eval_results = self.trainer.evaluate()
self.save_results()
```
- データセットを用いてモデルを学習させます。
- 評価データより分類性能を評価します。
- 後工程のために、モデルと指標を保存します。

## ステップ 2: LLM-as-a-judge 評価タスクの定義

本章では、ニュース分類ファインチューニング済みのモデルを評価するための、LLM を使用した評価パイプラインについて説明します。`scripts/evaluate/evaluate.py` の `TestGPT2Model` クラスは、定義したテストケースとニュースカテゴリ分類精度指標に従ってモデルを評価し、分類性能レポートを生成します。これにより、モデル間に一貫したベンチマークに従って比較できるため、モデルの精度向上に寄与します。

**1. 引数指定**

> `parse_arguments` 関数は、評価プロセスの必要な引数を処理します。さまざまなパラメータを任意に指定できるため、モデルのファインチューニングに柔軟的に変更で来ます。

Reviewer: ```suggestion
`parse_arguments` 関数は、評価プロセスの必要な引数を処理します。さまざまなパラメータを任意に指定できるため、モデルのファインチューニングに柔軟に対応できます。
```

knit plan active no ${plan_id}
```
`${plan_id}` を、非活性化したい Plan の一意の Id（例：`$train_plan_id`, `$evaluate_plan_id`）に置き換えてください。

### アップロードされたデータセットの削除:
Knitfabでアップロードされたデータセットを削除するには、関連付けられたアップロード Run を削除する必要があります。

**1. データセット Run Id の検索:**
- データセット一覧の取得：
  - 次のコマンドを実行して、タグ `project:news-classification` が付いたすべてのデータセットを取得します。
```bash
knit data find -t project:news-classification
```
- 実行例：
```json
{
  "knitId": "de3cec00-d723-40e3-b299-536bacdbbe44",
  "tags": [
    ...
    "project:news-classification",
    ...
  ],
  "upstream": {
    "mountpoint": {
      "path": "/upload",
      "tags": []
    },
    "run": {
      "runId": "596103be-a880-4a02-aea8-adfdbef1020b", 
      "status": "done",
      "updatedAt": "2025-01-27T10:14:28.237+00:00",
      "plan": {
        "planId": "cde9c9c3-a604-4057-9102-ed5f85d4afc7",
        "name": "knit#uploaded"
      }
    }
  },
  ...
}
```
- アップロード Run の特定：
  - 一覧の中から、`upstream.mountpoint.path` が `/upload` になっているエントリを特定します。
  - 該当する `upstream.run.runId` の値を抽出します。

**2. Run の削除:**
> [!Warning]
> 
> Run の削除は元に**不可逆的な**操作です。Runおよび関連するすべてのデータが完全に削除されます。
- 次のコマンドを実行します。
```bash
knit run rm ${run_id}
```
`${run_id}` を手順 1 で取得した `runId` に置き換えます。

### まとめ

> 本書では、Knitfab を活用して大規模言語モデル (LLM) ファインチューニングの自動化、および LLM-as-a-judge の性能評価を示していました。

Reviewer: ```suggestion
本書では、Knitfab を活用して大規模言語モデル (LLM) ファインチューニングの自動化、および LLM-as-a-judge の性能評価を示しました。
```

### 例
# Knitfab での LLM の微調整によるニュース分類

> 本書では、Knitfab を使用した一般大規模言語モデル (LLM) を微調整することにより、モデルがニュースといった特定の文書分類性能を向上することを示します。事例通して、Knitfab が微調整実験の入力と出力、および履歴の管理を自動化しながら、LLM-as-a-judge を用いて効率的に人間のような評価ができるようになります。

Reviewer: ```suggestion
本書では、Knitfab を使用した一般大規模言語モデル (LLM) をファインチューニングすることにより、モデルがニュースといった特定の文書分類性能を向上することを示します。この事例を通して、Knitfab がファインチューニング実験の入力と出力、および履歴の管理を自動化しながら、LLM-as-a-judge を用いて効率的に人間のような評価ができるようになります。
```

### 概要

> LLM の微調整というのは、特定のタスクに対して処理性能と判断性能を高めるために複数な超母数を調整する工程です。しかしながら、複雑な超母数によりLLM の微調整が早くとも手ごたえなくなります。そこで、 Knitfab は学習工程だけではなく学習の入出力、成果物まで自動管理を手伝うことで、さまざまな構成設定での実験が容易になります。さらに、革新的な評価方法である LLM-as-a-judge を用いることにより人間のような評価品質を維持できながらコストの削減も可能になります。本事例では、このニュース分類微調整から LLM-as-a-judge 評価の一連まで Knitfab 上の実装を体験し、機械学習自動管理プラットフォームとして Knitfab の強さと容易さを理解できます。

Reviewer: ```suggestion
LLM のファインチューニングは、特定のタスクに対して処理性能と判断性能を高めるために複数なハイパーパラメータを調整する工程です。しかしながら、複雑なハイパーパラメータによりLLM のファインチューニングの管理はすぐに難しいものになります。そこで、 Knitfab により学習工程にくわえて学習の入出力、成果物まで自動管理することで、さまざまな構成設定での実験が容易になります。さらに、革新的な評価方法である LLM-as-a-judge を用いることにより人間のような評価品質を維持しながらコストの削減も可能になります。本事例では、このニュース分類タスクのファインチューニングから LLM-as-a-judge 評価までの一連のタスクを Knitfab 上で実装することを体験し、リネージ自動管理プラットフォームとして Knitfab の強みと容易さをお伝えします。
```

In Japanese, Fine-Tuning and Hyperparameters are just "ファインチューニング" and "ハイパーパラメータ", respectively.
Your translation, 微調整 and 超母数 , may be TRUE, but not so natural.

After here, I do not comment about them, but replace.

ワークフローの主要なコンポーネントとプロセスの概要を図にて示します。

**コンポーネント:**
- __基底モデル:__ 微調整に使用される事前学習済み LLM。この例では、`GPT-2` を採用します。
- __学習データセット:__ モデルのニュース分類微調整に使用されるデータセット。`scikit-learn` の `20 Newsgroups` データセットを活用します。
- __微調整済みモデル:__ ニュース分類に特化した微調整済みモデル。
- __評価データセット:__ LLM-as-a-judge 評価に使用される `20 Newsgroups` データセットの `test` サブセット。
- __LLM 評価器:__ LLM-as-a-judge 評価のための LLM。
- __指標:__ LLLM-as-a-judge 評価のためのカスタム指標。

**プロセス:**
- __学習:__ ニュース記事を分類するために、学習データセットを用いて基底モデルを微調整する工程。
- __評価:__ カスタム指標と評価データセットを用いて、微調整済みモデルのパフォーマンス評価。

```mermaid
flowchart LR
    A((基底モデル)) --> C[学習]
    B((学習データセット)) --> C
    C --> D((微調整済みモデル))
    D --> G[評価]
    E((評価データセット)) --> G
    F((LLM 評価器)) --> G
    G --> H((指標)) 
```
### 前提条件

> 本事例を正常に動作するために、次の前提条件を満たしているかを確認してください。

Reviewer: ```suggestion
本事例を正常に動作させるために、次の前提条件を満たしているかを確認してください。
```

**必須:**
- **GPU:** 本例は GPU での実行を想定して設計されています。そのため、CPU での実行は非効率で処理時間が大幅に長くなります。
- **Knitfabのセットアップ:**
  - **プロダクション向けの場合:** [03.admin-guide](../../03.admin-guide/admin-guide-installation.ja.md)に記載されている手順に従って、インストールを実行します。
  - **実験向けの場合:** 簡易版については、[01.getting-started: Knitfab（簡易版） をローカル環境にインストールする](../../01.getting-started/getting-started.ja.md#Knitfab簡易版-をローカル環境にインストールする)に記載されている手順に従ってください。
- **Knit CLI:** [01.getting-started: CLI ツール: knit](../../01.getting-started/getting-started.ja.md#cli-ツール-knit)のインストール手順に従って、Knit CLI を設定します。
- **Knitコマンド初期化:** `knit init` を使用して Knit コマンドを初期化する方法については、[01.getting-started: knit コマンドの初期化](../../01.getting-started/getting-started.ja.md#knit-コマンドの初期化)を参照してください。
- **`Docker` インストール:** イメージをビルドして Knitfab にプッシュするために必要です。

> - **LLM 評価器**: LLM-as-a-Judge 評価には、LLM が評価器として設定する必要があります。次の 3 つのオプションの中から 1 つを選んで設定してください。

Reviewer: ```suggestion
- **LLM 評価器**: LLM-as-a-Judge 評価には、LLM を評価器として設定する必要があります。次の 3 つのオプションの中から 1 つを選んで設定してください。
```

  - **OpenAI**: https://docs.confident-ai.com/docs/metrics-introduction#using-openai ドキュメントの説明に従って OpenAI API キーを構成してください。
  - **カスタムモデル**: 独自のカスタムモデルを使用する場合、https://docs.confident-ai.com/docs/metrics-introduction#using-any-custom-llm に記載されている手順に従ってください。

**任意:**
- **`kubectl`**: LLM-as-a-Judge 評価に Ollama を使用したローカル LLM を選んだ場合は必須です。

### リポジトリ
本書で使用されるファイルとディレクトリは、GitHub の `knitfab-docs` リポジトリからをクローンできます。
```bash
git clone https://github.com/opst/knitfab-docs.git
```
クローンが完了したら、`04.examples/news-classification` ディレクトリ内にあるファイルを確認してください。
- **scripts:** このディレクトリには、モデルの学習と登録に必要なファイルが含まれています。
  - モデルの学習と評価のための Python スクリプト (`train.py`、`evaluate.py`)。
  - 必要なすべての Python パッケージを管理する `requirements.txt`。
  - Knitfab Plan の登録に使う Docker イメージを構築するための `Dockerfile`。
- **configs:** 学習と評価タスクの構成情報が含まれています。
- **plans:** Knitfab Plan の YAML ひな型が含まれています。

### タスク
- [ステップ 1: 学習と評価タスクの定義](#ステップ-1-学習と評価タスクの定義)
- [ステップ 2: Docker イメージをビルドして Knitfab にプッシュする](#ステップ-2-Docker-イメージをビルドして-Knitfab-にプッシュする)
- [ステップ 3: ニュース分類の微調整](#ステップ-3-ニュース分類の微調整)
- [ステップ 4: LLM-as-a-judge 評価](#ステップ-4-LLM-as-a-judge-評価)
- [ステップ 5: 片付け](#ステップ-5-片付け)

## ステップ 1: 学習と評価タスクの定義

### 1-1. 学習タスクの定義

> 本章では、`scripts/train/train.py` にニュース分類微調整のロジックを説明します。スクリプトは引数指定、モデル準備、データ処理、学習、と評価を構造化したパイプラインに従い、各コンポーネントの構成変更と最適化を容易にする設計を取っています。

Reviewer: ```suggestion
本章では、`scripts/train/train.py` にニュース分類微調整のロジックを説明します。スクリプトは引数指定、モデル準備、データ処理、学習と、評価を構造化したパイプラインに従い、各コンポーネントの構成変更と最適化を容易にする設計を取っています。
```

### 1. 引数指定

> `parse_arguments` 関数は、学習タスクに必要な引数を受け取ります。さまざまな母数を任意に指定できるため、モデルの微調整に柔軟的に変更できます。

Reviewer: ```suggestion
`parse_arguments` 関数は、学習タスクに必要な引数を受け取ります。さまざまなパラメータを任意に指定できるため、モデルの微調整に柔軟的に変更できます。
```

#### 主要な機能

> - `--config-file`: 構成ファイルのパス。指定された場合、既定引数を上書きます。

Reviewer: ```suggestion
- `--config-file`: 構成ファイルのパス。指定された場合、既定引数を上書きします。
```

    
    dataset = Dataset.from_pandas(df_cleaned)
    return dataset.map(tokenize, batched=True)
```
- 生データを `DataFrame` 形式に構造化します。
- データを `Transformer` で読み込んだ基底モデルの入力形式にトークン化します。

### 4. 学習超母数の設定
`setup_trainer` 関数は、適切な超母数とモデル構成を使用して `SFTTrainer` を初期化します。

#### 主要な機能
- `TrainingArguments` を使用して学習超母数を定義します。
- 微調整を効率化するために `LoraConfig` を構成します。
- `DataCollatorWithPadding` を使用してさまざまな文章を同じ固定長に変換します。

#### コード解析
```python
training_args = TrainingArguments(
    output_dir=self.args.save_to,
    logging_dir=f"{self.args.save_to}/logs",
    num_train_epochs=self.args.epochs,
    per_device_train_batch_size=1,
    per_device_eval_batch_size=1,
    gradient_accumulation_steps=2,
    learning_rate=self.args.learning_rate,
    weight_decay=0.05,
    max_grad_norm=0.3,
    fp16=True,
    bf16=False,
    warmup_ratio=0.1,
    lr_scheduler_type="cosine",
    logging_strategy="steps",
)
```
- バッチサイズ、学習率、重み減衰、勾配累積など、主要な学習超母数を定義します。
- 学習を効率化するために、混合精度 (`fp16`) を実装します。
- ウォームアップ比率 `0.1` 、とコサイン学習率スケジューラを使用します。
- 学習済みモデルやログファイルなど、プロジェクトの出力を整理するための専用ディレクトリを指定します。

```python
peft_config = LoraConfig(
    lora_alpha=16,
    lora_dropout=0.1,
    r=64,
    bias="none",
    task_type="SEQ_CLS",
)
```

> - 効率的な微調整のために `LoRA (Low-Rank Adaptation)` を実装します。

Reviewer: ```suggestion
- 効率的な微調整のために `LoRA (Low-Rank Adaptation)` を採用します。
```

> - 正則化とモデル分類性能の均衡を保つために、脱落値とランク値を設定します。

Reviewer: ```suggestion
- 正則化とモデル分類性能の均衡を保つために、ドロップアウトとランク値を設定します。
```

```python
collator = DataCollatorWithPadding(tokenizer=self.tokenizer, padding="max_length")
```
- 文章を同じ固定長に変換します。

```python
self.trainer = SFTTrainer(
    model=self.model,
    args=training_args,
    peft_config=peft_config,
    train_dataset=self.train_dataset,
    eval_dataset=self.eval_dataset,
    tokenizer=self.tokenizer,
    data_collator=collator,
)
```
- 構成済みのモデル、学習引数、データセット、およびトークナイザーを使用して `SFTTrainer` を初期化します。
- すべての要素を組み合わせて、分類微調整を効率的に実行します。

### 5. モデルの学習と評価

`run` 関数は、環境設定から結果の保存まで、学習プロセス全体を調整します。

#### 主要な機能
- データセットを読み込み、学習器を初期化します。
- 学習と評価を実装します。
- 学習済みモデルと評価指標を保存します。

#### コード解析
```python
self.trainer.train()
self.eval_results = self.trainer.evaluate()
self.save_results()
```
- データセットを用いてモデルを学習させます。
- 評価データより分類性能を評価します。
- 後工程のために、モデルと指標を保存します。

### 1-2. LLM-as-a-judge 評価

> 本章では、ニュース分類微調整済みのモデルを評価するために、LLM を使用した評価パイプラインを紹介します。`scripts/evaluate/evaluate.py` の `TestGPT2Model` クラスは、定義したテストケースとニュースカテゴリ分類精度指標に従ってモデルを評価し、分類性能レポートを生成します。これにより、モデル間に一貫した水準点で比較でき、モデルの精度向上を統一な方向性に従います。

Reviewer: ```suggestion
本章では、ニュース分類微調整済みのモデルを評価するための、LLM を使用した評価パイプラインについて説明します。`scripts/evaluate/evaluate.py` の `TestGPT2Model` クラスは、定義したテストケースとニュースカテゴリ分類精度指標に従ってモデルを評価し、分類性能レポートを生成します。これにより、モデル間に一貫したベンチマークに従って比較できるため、モデルの精度向上に寄与します。
```

### 1. 引数指定

> `parse_arguments` 関数は、評価プロセスの必要な引数を処理します。さまざまな母数を任意に指定できるため、モデルの微調整に柔軟的に変更で来ます。

Reviewer: ```suggestion
`parse_arguments` 関数は、評価プロセスの必要な引数を処理します。さまざまなパラメータを任意に指定できるため、モデルの微調整に柔軟的に変更で来ます。
```

#### 主要な機能

> - `--config-file`: 構成ファイルのパス。指定された場合、既定引数を上書きます。

Reviewer: ```suggestion
- `--config-file`: 構成ファイルのパス。指定された場合、既定引数を上書きします。
```

    expected_category = categories[category_idx]
    actual_output = map_category(result["label"], categories)
    test_case = LLMTestCase(
        input=text,
        actual_output=actual_output,
        expected_output=expected_category,
    )
    test_cases.append(test_case)
```
- 予測されたニュースカテゴリ (`actual_output`) を正解 (`expected_category`) と比較します。
- 評価用のテストケースとして保存します。

### 4. モデルパフォーマンスの評価

`TestGPT2Model` クラスは、`deepeval` という LLM-as-a-judge フレームワークを使用してモデルを評価します。

#### 主要な機能
- 分類性能の `CategoryPrecision` 指標を定義します。
- `deepeval.evaluate` を使用して評価を実施し、結果を保存します。

#### コード解析
```python
return GEval(
    name="CategoryPrecision",
    threshold=self.args.threshold,
    criteria="Category classifier precision - determine if the actual output matched the provided expected output",
    evaluation_steps=[
        "Award 1.0 points for exact match between actual and expected category",
        "Award 0.5 points for semantically related categories",
        "Award 0.0 points if categories are unrelated",
        "Minimum score after penalties is 0.0",
    ],
    evaluation_params=[LLMTestCaseParams.ACTUAL_OUTPUT, LLMTestCaseParams.EXPECTED_OUTPUT],
)
```
- 分類精度指標を定義します。

> - 不正確またはあいまいな予測に罰を与えます。

Reviewer: ```suggestion
- 不正確またはあいまいな予測にペナルティを与えます。
```

Yes, "Penalize" means "罰する". But it is not natural here.
"罰する" has a nuance of "by law", "by official rule" or "by boss". This case does not match such contexts.

### 2-1. Docker イメージをビルドする
### 1. `news-classification-train` イメージのビルド:
```bash
docker build -t news-classification-train:v1.0 \
             -f scripts/train/Dockerfile \
             ./scripts/train
```
`news-classification-train` イメージは、基底モデルの微調整を担います。

### 2. `news-classification-evaluate` イメージのビルド:

#### Dockerfile の変更:
`Dockerfile` 内のコマンドラインを変更して LLM-as-a-Judge の LLM 評価器を設定する必要があります。LLM 評価器の詳細設定については、[前提条件](#前提条件)に記載されているドキュメントを参照してください。

```docker
# LLM 評価器を設定するには、コマンド（base-url、api-keyなど）を変更してください。
RUN deepeval set-local-model --model-name=llama3.2 \
    --base-url="http://ollama:11434/v1/" \
    --api-key="ollama"
```

#### Dockerイメージのビルド:
```bash
docker build -t news-classification-evaluate:v1.0 \
             -f scripts/evaluate/Dockerfile \
             ./scripts/evaluate
```
`news-classification-evaluate` イメージは微調整済みモデルに対して LLM-as-a-judge 評価を行います。

### 2-2.（任意）Docker イメージの動作確認
> [!Note]
>
> ビルドしたイメージに自信がある場合は、Knitfab へのプッシュに進んでください。（[Docker イメージを Knitfab へプッシュする](#2-3-Docker-イメージを-Knitfab-へプッシュする)）

### 1. 微調整イメージの実行:
```bash
docker run --rm -it --gpus all \
    -v "$(pwd)/configs:/configs" \
    -v "$(pwd)/out:/out" \
    news-classification-train:v1.0
```

> このコマンドは、`news-classification-train:v1.0` イメージをロカール環境で実行します。

Reviewer: ```suggestion
このコマンドは、`news-classification-train:v1.0` イメージをローカル環境で実行します。
```

- `-v` フラグは、構成ファイル (`/configs`) と出力ディレクトリ (`/out`) をコンテナに関連付けます。
- これにより、イメージの動作確認ができ、微調整済みモデルが生成されます。

### 2. LLM-as-a-judge を使用したモデル評価:
> [!Caution]
> 

> > 次の手順から LLM-as-a-Judge の評価器して事前に LLM の設定が求められています。設定手順などについては、[前提条件](#前提条件)をご確認してください。

Reviewer: ```suggestion
> 以降は、 LLM-as-a-Judge の評価器として事前に LLM を設定しておく必要があります。設定手順などについては、[前提条件](#前提条件)をご確認してください。
```

>

> > 次の手順は、[ollama/ollama-setup.ja.md](ollama/ollama-setup.ja.md) 資料に説明されている Ollama を使用したローカル LLM の構成に基づいています。別の LLM 評価器を使用する場合、それに応じて手順を変更してください。

Reviewer: ```suggestion
> 次の手順は、[ollama/ollama-setup.ja.md](ollama/ollama-setup.ja.md) に説明されている Ollama を使用したローカル LLM の構成に基づいています。別の LLM 評価器を使用する場合、それに応じて手順を変更してください。
```

- イメージは定義されたカスタム指標と `20 Newsgroups` データセットの `test` サブセットを使用して、微調整済みモデルを評価します。
- 評価指標は、`out` ディレクトリに `deepeval-result.json` の JSON ファイルとして保存されます。

### 3. パフォーマンス分析:

`deepeval-result.json` ファイルを確認して、定義されたカスタム指標に基づいてモデルの性能を分析します。カスタム指標が期待要件を適切に反映しているかどうかを検討し、必要に応じて調整・再定義してください。

### 2-3. Docker イメージを Knitfab へプッシュする

### 1. レジストリ URI を用いてイメージのタグ付け:

イメージを Knitfab レジストリにプッシュする前に、正しいレジストリ URI でタグ付けする必要があります。これにより、Docker はプッシュ操作の目的のレジストリを識別できます。
```bash
docker tag ${docker_image} ${registry_uri}/${docker_image}
```
プレースホルダを値で置き換えてください:

- `${docker_image}` をビルドされた各イメージの名前（例：`news-classification-train:v1.0`, `news-classification-evaluate:v1.0`）に置き換えてください。
- `${registry_uri}` を、Knitfab レジストリの実際の URI（例：`192.0.2.1:30503`）に置き換えてください。

### 2. イメージをKnitfabレジストリへプッシュ:

次に、タグ付きのイメージを Knitfab レジストリにプッシュします。
```bash
docker push ${registry_uri}/${docker_image}
```
`${docker_image}` を、前の手順でタグ付けした各イメージの名前（レジストリ URI を含む）に置き換えてください。

## ステップ 3: ニュース分類の微調整
本章では、Knitfab 上 LLM のニュース分類微調整を行います。

### 1. YAML ひな型の生成:

YAML ひな型を生成するには、2 つの選択肢があります。

- 選択肢①：空のひな型を作成する
```bash
knit plan template --scratch > ./plans/news-classification-train.v1.0.yaml
```

> これにより、`./plans` ディレクトリに`news-classification-train.v1.0.yaml` という名前で新しい空の YAML ひな型が作成されます。ひな型には必要な構成要素を追記します。

Reviewer: ```suggestion
これにより、`./plans` ディレクトリに`news-classification-train.v1.0.yaml` という名前で新しい空の YAML ひな型が作成されます。ひな型に、後述する必要な構成要素を追記します。
```

- 選択肢②：Docker イメージからひな型を生成する
```bash
docker save ${registry_uri}/news-classification-train:v1.0 | \
    knit plan template > ./plans/news-classification-train.v1.0.yaml
```
このコマンドは、Docker イメージ `news-classification-train:v1.0` に基づいて YAML ひな型を生成します。

\* `${registry_uri}` を 実際の Knitfab のレジストリ URI に置き換えてください。

> \* コマンドが Docker イメージを解析して、ひな型の一部分が記入済みです。

Reviewer: ```suggestion
\* コマンドが Docker イメージを解析して、ひな型の一部を記入したものが生成されます。
```

### 2. YAML ひな型の追記:
- 重要な追記点:
  - `image`:
    Knitfab Kubernetes クラスタがローカルレジストリを使用している場合は、`image` 項目の`registry_uri` を `localhost` に置き換えてください。
    
    例:
    ```YAML
    # 192.0.2.1 を
    image: "192.0.2.1:30503/news-classification-train:v1.0"
    # localhost に置き換える
    image: "localhost:30503/news-classification-train:v1.0"
    ```

>     - これにより、Knitfab はローカルレジストリからイメージをプルするという意味合いになります。

Reviewer: ```suggestion
    - これにより、Knitfab は、Knitfab 内のローカルレジストリからイメージをプルするという意味合いになります。
```

      - path: "/configs"
        tags:
          - "type:config"
          - "project:news-classification"
    
    outputs:
      - path: "/out/model"
        tags:
          - "type:model"
          - "project:news-classification"
    
    log:
      tags:
        - "type:log"
        - "project:news-classification"
    ```
      - これにより、Knitfab 内での入力、出力とログの整理と検索が容易になります。
  
  - 計算資源: 
    学習プロセスに必要な資源（例：CPU、メモリ、GPU）を定義します。

    ```YAML
    resources:
      cpu: 1
      memory: 10Gi
      nvidia.com/gpu: "1"
    ```

- その他の重要な考慮事項：
  変更した YAML ひな型が正しい構造と構文に準拠していることを再確認してください。必要に応じてクローンした Git リポジトリの `/plans` ディレクトリにある YAML ファイルを参照してください。

### 3. YAML ひな型の登録:
```bash
train_plan=$(knit plan apply ./plans/news-classification-train.v1.0.yaml)
```
このコマンドは、YAML ひな型を Knitfab API に送信し、定義した構成要素に沿って新しい Plan を作成します。

`train_plan` 変数にはコマンドの出力として作成された Plan の詳細が格納されています。

### 4. Plan Id の抽出:
```bash
train_plan_id=$(echo "$train_plan" | jq -r '.planId')
```
これで、Plan の詳細から Plan の一意の Id を抽出できます。

### 5. 構成ファイルを Knitfab へプッシュ:
```bash
knit data push -t type:config \
               -t project:news-classification \
               -n ./configs
```
このコマンドは、`./configs`にある構成ファイルを Knitfab にプッシュします。

> `-t` フラグは、タグ（`type:config`, `project:news-classification`）を追加することで、Knitfab が構成ファイルを Plan に関連付けられるようになります。

Reviewer: ```suggestion
### 5. 設定ファイルを Knitfab へプッシュ:
```bash
knit data push -t type:config \
               -t project:news-classification \
               -n ./configs
```
このコマンドは、`./configs`にある設定ファイルを Knitfab にプッシュします。

`-t` フラグによってタグ（`type:config`, `project:news-classification`）を追加することで、Knitfab が設定ファイルを Plan に関連付けられるようになります。
```

### 6. 実行状況の確認:

YAML ひな型を登録した後、Knitfab は学習 Plan を実行するための Run が開始します。次のコマンドを使用して、Runの実行状況を監視できます。
```bash
knit run find -p $train_plan_id
```
このコマンドは、指定された Plan Id に関連付けられた学習 Run を表示します。コマンドを定期的に実行し、`status` が `done` に変わるまで待ちます。`done` に変わると Run が正常終了になります。

### 7. モデル情報の取得:

学習 Run が正常に終了したら、生成されたモデルに関する情報を取得できます。
- Run 情報の取得:
```bash
train_run=$(knit run find -p $train_plan_id)
```
- Run 出力の抽出:
```bash
train_outputs=$(echo "$train_run" | jq -r '.[-1].outputs')
```
- モデル Knit Id の取得:
```bash
train_model_knit_id=$(echo "$train_outputs" | jq -r '.[0].knitId')
```

### 8.（任意）Run ログの確認:

学習中にログを確認したい場合は、次のコマンドを使用できます。
- Run Id の取得:
```bash
train_run_id=$(echo "$train_run" | jq -r '.[-1].runId')
```
- Run ログの表示:
```bash
knit run show --log $train_run_id
```

### 9.（任意）モデルの保存:

学習済みのモデルをローカルに保存したい場合は、次のコマンドを使用して Knitfab からダウンロードできます。
```bash
knit data pull -x $train_model_knit_id ./out
```
このコマンドは、学習済みモデルを Knitfab からダウンロードし、`./out` ディレクトリに保存します。

## ステップ 4: LLM-as-a-judge 評価
> [!Caution]
> 

> > 次の手順から LLM-as-a-Judge の評価器して事前に LLM の設定が求められています。設定手順などについては、[前提条件](#前提条件)を再度ご確認してください。

Reviewer: ```suggestion
> 以降の手順では LLM-as-a-Judge の評価器として事前に LLM を設定する必要があります。設定手順などについては、[前提条件](#前提条件)を再度ご確認ください。
```

```bash
knit plan template --scratch > ./plans/news-classification-evaluate.v1.0.yaml
```
- 選択肢②：Docker イメージからひな型を生成する
```bash
docker save ${registry_uri}/news-classification-evaluate:v1.0 | \
    knit plan template > ./plans/news-classification-evaluate.v1.0.yaml
```
\* `${registry_uri}` を 実際の Knitfab のレジストリ URI に置き換えてください。

### 2. YAML ひな型の追記:
- 重要な追記点:
  - `image`:
    Knitfab Kubernetes クラスタがローカルレジストリを使用している場合は、`image` 項目の`registry_uri` を `localhost` に置き換えてください。

    例:
    ```YAML
    # Replace 192.0.2.1
    image: "192.0.2.1:30503/news-classification-evaluate:v1.0"
    # With localhost
    image: "localhost:30503/news-classification-evaluate:v1.0"
    ```
  - `inputs`, `outputs`, `log`:
    以下のように関連タグを追加してください。
    
    ```YAML
    inputs:
      - path: "/configs"
        tags:
          - "type:config"
          - "project:news-classification"
      
      - path: "/in"
        tags:
          - "type:model"
          - "project:news-classification"

    outputs:
      - path: "/out"
        tags:
          - "type:metrics"
          - "project:news-classification"

    log:
      tags:
        - "type:log"
        - "project:news-classification"
    ```
> [!Note]
> 

> > 評価タスクが学習タスクの出力モデルを動的に認識できるように、`in/model` は構成ファイル内で定義せず、Plan YAML ひな型内で明示的に追加します。

Reviewer: ```suggestion
> 評価タスクが学習タスクの出力モデルを動的に認識できるように、`in/model` は設定ファイル内で定義せず、Plan YAML ひな型内で明示的に追加します。
```

  - これにより、Knitfab 内での入力、出力とログの整理と検索が容易になります。
  
  - 計算資源: 
    学習プロセスに必要な資源（例：CPU、メモリ、GPU）を定義します。

    ```YAML
    resources:
      cpu: 1
      memory: 10Gi
      nvidia.com/gpu: "1"
    ```

- その他の重要な考慮事項：
  変更した YAML ひな型が正しい構造と構文に準拠していることを再確認してください。必要に応じてクローンした Git リポジトリの `/plans` ディレクトリにある YAML ファイルを参照してください。

### 3. YAML ひな型の登録:
```bash
evaluate_plan=$(knit plan apply ./plans/news-classification-evaluate.v1.0.yaml)
```
- `evaluate_plan` 変数にはコマンドの出力として作成された Plan の詳細が格納されています。

> - 構成ファイルは前章にて Knitfab にプッシュされているため、`evaluate_plan` の基に Run が自動的に生成され実行されます。

Reviewer: ```suggestion
- 設定ファイルは前章にて Knitfab にプッシュされているため、`evaluate_plan` の基に Run が自動的に生成され実行されます。
```

### 4. Plan Id の抽出:
```bash
evaluate_plan_id=$(echo "$evaluate_plan" | jq -r '.planId')
```
これで、Plan の詳細から Plan の一意の Id を抽出できます。

### 5. 実行状況の確認:
```bash
knit run find -p $evaluate_plan_id
```
`status` が `done` に変わるまで待ちます。`done` に変わると Run が正常終了になります。

### 6.（任意）評価指標の取得:

- Run 情報の取得:
```bash
evaluate_run=$(knit run find -p $evaluate_plan_id)
```
- Run出力の抽出:
```bash
evaluate_outputs=$(echo "$evaluate_run" | jq -r '.[-1].outputs')
```
- 評価指標レポートの Knit Id の取得:
```bash
evaluate_metrics_knit_id=$(echo "$evaluate_outputs" | jq -r '.[0].knitId')
```

### 7.（任意）Run ログの確認:
- Run Id の取得:
```bash
evaluate_run_id=$(echo "$evaluate_run" | jq -r '.[-1].runId')
```
- Run ログの表示:
```bash
knit run show --log $evaluate_run_id
```

### 8.（任意）指標レポートのダウンロード:
```bash
knit data pull -x $evaluate_metrics_knit_id ./out
```
このコマンドは、評価指標レポートを Knitfab からダウンロードし、`./out` ディレクトリに保存します。

## ステップ 5： 片付け
### 5-1. Run の 削除：
> [!Caution]
> 
> - Run が**すでに停止**していて、**下流に他の Run がない**場合に限り、Run を削除できます。
> - Run を削除してしまうと、その Run によって生成された**すべてのデータ、モデル、指標レポートも完全に削除**されます。

> [!Warning]
>

> > Run の 削除は**非可逆**の操作です。

Reviewer: ```suggestion
> Run の 削除は**不可逆的な** 操作です。
```

### 5-3. アップロードされたデータセットの削除：
Knitfabでアップロードされたデータセットを削除するには、関連付けられたアップロード Run を削除する必要があります。

### 1. データセット Run Id の検索:
- データセット一覧の取得：
  - 次のコマンドを実行して、タグ `project:news-classification` が付いたすべてのデータセットを取得します。
```bash
knit data find -t project:news-classification
```
- 実行例：
```json
{
  "knitId": "de3cec00-d723-40e3-b299-536bacdbbe44",
  "tags": [
    ...
    "project:news-classification",
    ...
  ],
  "upstream": {
    "mountpoint": {
      "path": "/upload",
      "tags": []
    },
    "run": {
      "runId": "596103be-a880-4a02-aea8-adfdbef1020b", 
      "status": "done",
      "updatedAt": "2025-01-27T10:14:28.237+00:00",
      "plan": {
        "planId": "cde9c9c3-a604-4057-9102-ed5f85d4afc7",
        "name": "knit#uploaded"
      }
    }
  },
  ...
}
```
- アップロード Run の特定：
  - 一覧の中から、`upstream.mountpoint.path` が `/upload` になっているエントリを特定します。
  - 該当する `upstream.run.runId` の値を抽出します。

### 2. Run の削除:
> [!Warning]
> 
> Run の削除は元に**不可逆的な**操作です。Runおよび関連するすべてのデータが完全に削除されます。
- 次のコマンドを実行します。
```bash
knit run rm ${run_id}
```
`${run_id}` を手順 1 で取得した `runId` に置き換えます。

### まとめ
本書では、Knitfab を活用して大規模言語モデル (LLM) 微調整の自動化、および LLM-as-a-judge の性能評価を示していました。

> - **Knitfab による自動化された LLM 微調整**: Knitfab は学習工程から学習成果物の管理まで自動化することで、さまざまな構成に対して実験を容易にし、進捗状況の明確な記録を維持できるようになります。これにより、手動工程が減り、LLM の学習と微調整が速やかになります。

Reviewer: ```suggestion
- **Knitfab による自動化された LLM 微調整**: Knitfab は学習工程から学習成果物の管理まで自動化することで、さまざまな構成に対して実験を容易にし、進捗状況の明確な記録を維持できるようになります。これにより、手動工程が減り、LLM の学習と微調整が容易になります。
```

> - **LLM-as-a-Judge による効率的な LLM 評価**: LLM を評価器として採用することにより、人間と同等評価品質を維持できながら、費用対効果を高め、時間を節約できるようになります。

Reviewer: ```suggestion
- **LLM-as-a-Judge による効率的な LLM 評価**: LLM を評価器として採用することにより、人間と同等程度の評価品質を維持しながら、費用対効果を高め、時間を節約できるようになります。
```

**任意:**
- **`kubectl`**: LLM-as-a-Judge 評価に Ollama を使用したローカル LLM を選んだ場合は必須です。

### リポジトリ
本書で使用されるファイルとディレクトリは、GitHub の `knitfab-docs` リポジトリからをクローンできます。
```bash
git clone https://github.com/opst/knitfab-docs.git
```
クローンが完了したら、`04.examples/news-classification` ディレクトリ内にあるファイルを確認してください。
- **scripts:** このディレクトリには、モデルの学習と登録に必要なファイルが含まれています。
  - モデルの学習と評価のための Python スクリプト (`train.py`、`evaluate.py`)。
  - 必要なすべての Python パッケージを管理する `requirements.txt`。
  - Knitfab Plan の登録に使う Docker イメージを構築するための `Dockerfile`。
- **configs:** 学習と評価タスクの構成情報が含まれています。
- **plans:** Knitfab Plan の YAML ひな型が含まれています。

### タスク
- [ステップ 1: 学習と評価タスクの定義](#ステップ-1-学習と評価タスクの定義)
- [ステップ 2: Docker イメージをビルドして Knitfab にプッシュする](#ステップ-2-Docker-イメージをビルドして-Knitfab-にプッシュする)
- [ステップ 3: ニュース分類のファインチューニング](#ステップ-3-ニュース分類のファインチューニング)
- [ステップ 4: LLM-as-a-judge 評価](#ステップ-4-LLM-as-a-judge-評価)
- [ステップ 5: 片付け](#ステップ-5-片付け)

## ステップ 1: 学習と評価タスクの定義
---

### 学習タスクの定義:

本章では、`scripts/train/train.py` にニュース分類ファインチューニングのロジックを説明します。スクリプトは引数指定、モデル準備、データ処理、学習と、評価を構造化したパイプラインに従い、各コンポーネントの構成変更と最適化を容易にする設計を取っています。

#### 1. 引数指定

> `parse_arguments` 関数は、学習タスクに必要な引数を受け取ります。さまざまなパラメータを任意に指定できるため、モデルのファインチューニングに柔軟的に変更できます。

Reviewer: ```suggestion
`parse_arguments` 関数は、学習タスクに必要な引数を受け取ります。さまざまなパラメータを任意に指定できるため、モデルのファインチューニングに柔軟に変更できます。
```

##### 主要な機能
- `--config-file`: 設定ファイルのパス。指定された場合、既定引数を上書きします。
- `--save-to`: ファインチューニング済みモデルとログが保存されるディレクトリです。既定値は `./out` です。
- `--base-model`: ファインチューニングに使用する基底モデルを指定します。既定値は `openai-community/gpt2` です。
- `--device`: 学習が実施されるハードウェア (`cuda` または `cpu`) を定義します。既定値は `auto` です。
- `--epochs`: 学習回数. 既定値は `1` です。
- `--learning-rate`: 学習率を指定します。既定値は `2e-5` です。

> 設定ファイルとコマンドラインの両方で因数を指定した場合、設定ファイルの設定内容が優先されます。

Reviewer: ```suggestion
設定ファイルとコマンドラインの両方で引数を指定した場合、設定ファイルの設定内容が優先されます。
```

typo?

In Japanese, "引数" is read as "ひきすう"。

      - path: "/configs"
        tags:
          - "type:config"
          - "project:news-classification"
    
    outputs:
      - path: "/out/model"
        tags:
          - "type:model"
          - "project:news-classification"
    
    log:
      tags:
        - "type:log"
        - "project:news-classification"
    ```
      - これにより、Knitfab 内での入力、出力とログの整理と検索が容易になります。
  
  - 計算資源: 
    学習プロセスに必要な資源（例：CPU、メモリ、GPU）を定義します。

    ```YAML
    resources:
      cpu: 1
      memory: 10Gi
      nvidia.com/gpu: "1"
    ```

- その他の重要な考慮事項：
  変更した YAML ひな型が正しい構造と構文に準拠していることを再確認してください。必要に応じてクローンした Git リポジトリの `/plans` ディレクトリにある YAML ファイルを参照してください。

#### 3. YAML ひな型の登録:
```bash
train_plan=$(knit plan apply ./plans/news-classification-train.v1.0.yaml)
```
このコマンドは、YAML ひな型を Knitfab API に送信し、定義した構成要素に沿って新しい Plan を作成します。

`train_plan` 変数にはコマンドの出力として作成された Plan の詳細が格納されています。

#### 4. Plan Id の抽出:
```bash
train_plan_id=$(echo "$train_plan" | jq -r '.planId')
```
これで、Plan の詳細から Plan の一意の Id を抽出できます。

#### 5. 設定ファイルを Knitfab へプッシュ:
```bash
knit data push -t type:config \
               -t project:news-classification \
               -n ./configs
```
このコマンドは、`./configs`にある設定ファイルを Knitfab にプッシュします。

> `-t` フラグは、タグ（`type:config`, `project:news-classification`）を追加することで、Knitfab が設定設定ファイルを Plan に関連付けられるようになります。

Reviewer: ```suggestion
`-t` フラグは、タグ（`type:config`, `project:news-classification`）を追加することで、Knitfab 設定ファイルを Plan に関連付けられるようになります。
```

typo?

    # With localhost
    image: "localhost:30503/news-classification-evaluate:v1.0"
    ```
  - `inputs`, `outputs`, `log`:
    以下のように関連タグを追加してください。
    
    ```YAML
    inputs:
      - path: "/configs"
        tags:
          - "type:config"
          - "project:news-classification"
      
      - path: "/in"
        tags:
          - "type:model"
          - "project:news-classification"

    outputs:
      - path: "/out"
        tags:
          - "type:metrics"
          - "project:news-classification"

    log:
      tags:
        - "type:log"
        - "project:news-classification"
    ```
> [!Note]
> 
> 評価タスクが学習タスクの出力モデルを動的に認識できるように、`in/model` は設定ファイル内で定義せず、Plan YAML ひな型内で明示的に追加します。

  - これにより、Knitfab 内での入力、出力とログの整理と検索が容易になります。
  
  - 計算資源: 
    学習プロセスに必要な資源（例：CPU、メモリ、GPU）を定義します。

    ```YAML
    resources:
      cpu: 1
      memory: 10Gi
      nvidia.com/gpu: "1"
    ```

- その他の重要な考慮事項：
  変更した YAML ひな型が正しい構造と構文に準拠していることを再確認してください。必要に応じてクローンした Git リポジトリの `/plans` ディレクトリにある YAML ファイルを参照してください。

#### 3. YAML ひな型の登録:
```bash
evaluate_plan=$(knit plan apply ./plans/news-classification-evaluate.v1.0.yaml)
```
- `evaluate_plan` 変数にはコマンドの出力として作成された Plan の詳細が格納されています。

> - 設定設定ファイルは前章にて Knitfab にプッシュされているため、`evaluate_plan` の基に Run が自動的に生成され実行されます。

Reviewer: ```suggestion
- 設定ファイルは前章にて Knitfab にプッシュされているため、`evaluate_plan` の基に Run が自動的に生成され実行されます。
```
